src/Http/Client/Cookie/Cookie.php

```diff
@@ -0,0 +1,387 @@
+ *
+ * Licensed under The MIT License
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.5.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Http\Client\Cookie;
+
+use Cake\Utility\Hash;
+use Cake\Utility\Security;
+use DateTimeInterface;
+use InvalidArgumentException;
+use RuntimeException;
+
+class Cookie implements CookieInterface
+{
+
+    use CookieCryptTrait;
+
+    /**
+     * Cookie name
+     *
+     * @var string
+     */
+    protected $name;
+
+    /**
+     * Raw Cookie value
+     */
+    protected $value;
+
+    /**
+     * Cookie data
+     *
+     * If the raw cookie data was a serialized array and was expanded
+     * this property will keep the data.
+     *
+     * @var array
+     */
+    protected $data;
+
+    /**
+     * @var bool
+     */
+    protected $isExpanded = false;
+
+    /**
+     * HTTP only
+     *
+     * @var bool
+     */
+    protected $expiresAt = 0;
+
+    /**
+     * Path
+     *
+     * @var string
+     */
+    protected $path = '';
+
+    /**
+     * Domain
+     *
+     * @var string
+     */
+    protected $domain = '';
+
+    /**
+     * Secure
+     *
+     * @var bool
+     */
+    protected $secure = false;
+
+    /**
+     * HTTP only
+     *
+     * @var bool
+     */
+    protected $httpOnly;
+
+    /**
+     * The key for en- end decrypting the cookie
+     *
+     * @var string
+     */
+    protected $encryptionKey;
+
+    /**
+     * Constructor
+     *
+     * @param string $name Cookie name
+     * @param string $value Value of the cookie
+     */
+    public function __construct($name, $value)
+    {
+        $this->validateName($name);
+        $this->setName($name);
+        $this->setValue($value);
+    }
+
+    /**
+     * Returns a header value as string
+     *
+     * @return string
+     */
+    public function toHeaderValue()
+    {
+        $headerValue = sprintf('%s=%s', $this->name, urlencode($this->value));
+        if ($this->expiresAt !== 0) {
+            $headerValue .= sprintf(
+                '; expires=%s',
+                gmdate('D, d-M-Y H:i:s T', $this->expiresAt)
+            );
+        }
+        if (empty($this->path) === false) {
+            $headerValue .= sprintf('; path=%s', $this->path);
+        }
+        if (empty($this->domain) === false) {
+            $headerValue .= sprintf('; domain=%s', $this->domain);
+        }
+        if ($this->secure) {
+            $headerValue .= '; secure';
+        }
+        if ($this->httpOnly) {
+            $headerValue .= '; httponly';
+        }
+
+        return $headerValue;
+    }
+
+    /**
+     * Sets the cookie name
+     *
+     * @param string $name Name of the cookie
+     * @return $this
+     */
+    public function setName($name)
```

Comment:
I don't mind doing it but what is the exact benefit in this use case? For example by reading [this article](https://www.linkedin.com/pulse/20140528113353-16837833-6-benefits-of-programming-with-immutable-objects-in-java) (if it's correct)  

>Do not change the state of the objects in any methods of the class`

 we would need to make all property setters return a new copy of the object because if we make one immutable we should make all other so as well?